File: edp_java/src/main/Cook.java
package main;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Semaphore;

public class Cook extends Thread {

    // Constants (Time in milliseconds)
    private static final int COOKING_TIME = 6000;
    private static final int COFFEE_BREAK_TIME = 3000;

    // Static
    private static final Cook[] CHEFS = java.util.Arrays.stream(Name.COOK_NAMES)
            .map(n -> new Cook(n.toString())).toArray(Cook[]::new);

    public static Cook[] list() {
        return CHEFS;
    }

    protected static final BlockingQueue<Order> ORDER_QUEUE = new java.util.concurrent.ArrayBlockingQueue<>(1);

    private static final Semaphore cookSemaphore = Waiter.cookSemaphore;

    // Instance
    private final String name;
    private int mealsPrepared = 0;
    private boolean isOnBreak = false;

    public Cook(String name) {
        this.name = name;
    }

    public String name() {
        return name;
    }

    public boolean isOnBreak() {
        return isOnBreak;
    }

    @Override
    public void run() {
        while (!Thread.currentThread().isInterrupted()) {
            try {
                System.out.println(String.format("Chef %s is waiting for an order.", name));
                Order order = ORDER_QUEUE.take();
                if (order != null) {
                    cook(order);
                    mealsPrepared++;
                    if (mealsPrepared % 4 == 0) {
                        takeCoffeeBreak();
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    private void cook(Order order) throws InterruptedException {
        System.out.println(String.format("Chef %s is cooking the %s for Philosopher %s.", name, order.mealString(),
                order.philosopher().name()));
        Thread.sleep(COOKING_TIME);
        Counter.placeCompletedMeal(order);
        System.out.println(String.format("Chef %s has finished cooking the %s for Philosopher %s.", name,
                order.mealString(), order.philosopher().name()));
        cookSemaphore.release();
    }

    private void takeCoffeeBreak() throws InterruptedException {
        isOnBreak = true;
        System.out.println(String.format("Chef %s has returned from a coffee break.", name));
        Thread.sleep(COFFEE_BREAK_TIME);
        isOnBreak = false;
    }
}

File: edp_go/cook.go
package main

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type CookManager struct {
	cooks      []*Cook
	orderQueue chan *Order
}

var cookManager = &CookManager{
	orderQueue: make(chan *Order, 1),
}

func init() {
	cookManager.cooks = make([]*Cook, len(COOK_NAMES))
	for i, name := range COOK_NAMES {
		cookManager.cooks[i] = NewCook(name)
	}
}

func (cm *CookManager) List() []*Cook {
	result := make([]*Cook, len(cm.cooks))
	copy(result, cm.cooks)
	return result
}

type Cook struct {
	name          string
	mealsPrepared int
	isOnBreak     bool
	mutex         sync.RWMutex
	ctx           context.Context
	cancel        context.CancelFunc
	wg            sync.WaitGroup
}

func NewCook(name Name) *Cook {
	ctx, cancel := context.WithCancel(context.Background())
	return &Cook{name: string(name), ctx: ctx, cancel: cancel}
}

func (c *Cook) Name() string { return c.name }
func (c *Cook) IsOnBreak() bool {
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	return c.isOnBreak
}

func (c *Cook) Run() {
	defer c.wg.Done()
	for {
		select {
		case <-c.ctx.Done():
			return
		default:
			fmt.Printf("Chef %s is waiting for an order.\n", c.name)
			select {
			case order := <-cookManager.orderQueue:
				if order != nil {
					c.cook(order)
					c.mutex.Lock()
					c.mealsPrepared++
					if c.mealsPrepared%4 == 0 {
						c.mutex.Unlock()
						c.takeCoffeeBreak()
						c.mutex.Lock()
					}
					c.mutex.Unlock()
				}
			case <-c.ctx.Done():
				return
			}
		}
	}
}

func (c *Cook) cook(order *Order) {
	fmt.Printf("Chef %s is cooking the %s for Philosopher %s.\n", c.name, order.MealString(),
		order.Philosopher().Name())
	time.Sleep(COOKING_TIME)
	counter.PlaceCompletedMeal(order)
	fmt.Printf("Chef %s has finished cooking the %s for Philosopher %s.\n", c.name,
		order.MealString(), order.Philosopher().Name())
	
	select {
	case waiterManager.cookSemaphore <- struct{}{}:
	default:
	}
}

func (c *Cook) takeCoffeeBreak() {
	c.mutex.Lock()
	c.isOnBreak = true
	c.mutex.Unlock()
	
	fmt.Printf("Chef %s is taking a coffee break.\n", c.name)
	time.Sleep(COFFEE_BREAK_TIME)
	
	c.mutex.Lock()
	c.isOnBreak = false
	c.mutex.Unlock()
	fmt.Printf("Chef %s has returned from a coffee break.\n", c.name)
} 